streams/src/main/java/org/apache/kafka/streams/errors/StreamsUncaughtExceptionHandler.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.errors;
+
+public interface StreamsUncaughtExceptionHandler {
+    /**
+     * Inspect a record and the exception received.
+     * @param exception the actual exception
+     */
+    StreamsUncaughtExceptionHandler.StreamsUncaughtExceptionHandlerResponse handle(final Exception exception);
```

Comment:
Let's tweak this API to Throwable:
```suggestion
    StreamsUncaughtExceptionHandler.StreamsUncaughtExceptionHandlerResponse handle(final Throwable exception);
```

Here's a good explanation of why: https://stackoverflow.com/questions/2274102/difference-between-using-throwable-and-exception-in-a-try-catch

The benefit is that we could handle `Error`s as well as `Exception`s. However, this comes with the obligation that we should not continue to use the thread after an Error occurs. I think we can deal with this restriction reasonably.

Reply:
Hmmm that make sense mostly. However it seems to handle the application shutdown case we need to the keep the thread alive to trigger a rebalance. Could we somehow remove the option of shutdown for some cases?

Reply:
That's a good question. Maybe we could just document that that option is an unsuitable response to an Error, and also log an `ERROR` message if you select it in response to an Error. It's not _always_ bad to ignore an Error, but it usually is. We can leave it to users to decide what they want to do.

Reply:
That seems to be a fine solution. It will still attempt the shutdown but may fail. As long as we warn I guess it will work